rfcs/convergence/props-typing-improvements.md

```diff
@@ -0,0 +1,164 @@
+
+```typescript
+export interface WidgetProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  icon?: ShorthandProps<IconProps>;
+  text?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;
+
+  // This prop has a default defined in useWidget:
+  propWithDefaultValue?: string;
+
+  // This prop does not have a default value in useWidget (undefined by default)
+  anotherProp?: boolean;
+}
+```
+
+**Problem:** The `WidgeState` type is handwritten, requires repeating the types for slot props and any props that have defaults.
+
+```typescript
+export interface WidgetState extends WidgetProps {
+  icon?: ObjectShorthandProps<IconProps>;
+  text: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>>;
+  propWithDefaultValue: string;
+}
+```
+
+**Problem:** there's no checking that `widgetShorthandProps` actually references shorthand props (it could include any random HTML prop as well, like `'name'`).
+
+```typescript
+export const widgetShorthandProps: (keyof WidgetProps)[] = ['icon', 'text'];
+```
+
+**Problem:** there's no type checking for the default values provided to `mergeProps`, and no checking that props that require a default value are provided. _(Can you spot the bug below?)_
+
+```typescript
+export const useWidget = (props: WidgetProps, ref: React.Ref<HTMLElement>, defaultProps?: WidgetProps): WidgetState => {
+  const state = mergeProps(
+    {
+      as: 'div',
+      test: { as: 'span' },
+      propWithDefaultValue: 'hello world',
+    },
+    defaultProps,
+    resolveShorthandProps(props, widgetShorthandProps),
+  );
+
+  return state;
+};
+```
+
+**Problem:** in hooks like `useWidgetStyles`, there's either an implicit assumption about default props, or unnecessary checking is performed:
+
+```typescript
+// Unchecked assumption that the value is not undefined:
+[s => s.propWithDefaultValue!.contains('hello'), {/* ... */}],
+// Unnecessary undefined check:
+[s => s.propWithDefaultValue && s.propWithDefaultValue.contains('hello'), {/* ... */}],
+```
+
+## Detailed Design or Proposal
+
+The fix would be to:
+
+1. Add some helper types
+   - `ComponentState`: assists in creating the `State` type from the `Props` type: the dev simply lists the names of the props that are slot props, and the props that have default values.
+   - `ComponentShorthandProps`: a list of the shorthand props from the state.
+2. Update `mergeProps` (`makeMergeProps`) to require that the first argument is of type `TState`.
+   - This will enforce that all props with defaults are provided and are of the correct type.
+3. Declare the `widgetShorthandProps` array "`as const`" and put it in the .types.ts file.
+   - This allows it to be used to declare the type used for the shorthand prop names.
+
+### Usage example
+
+`Widget.types.ts`:
+
+```typescript
+export const widgetShorthandProps = ['icon', 'text'] as const;
+
+export type WidgetState = ComponentState<

```

Comment:
Internally, `useWidget` calls `mergeProps` to get its return object, which is a function that could merge arbitrary objects together. Defining the return type of `mergeProps` using generic types based on its input is Hard/impossible because of how the merging actually takes place (some props are deep-merged, some are shallow-merged). So defining `WidgetState` in the way you mentioned wouldn't really work.

So instead, we define this `WidgetState` type that defines the expected result of `mergeProps`. We know that if we start with a `WidgetState` and only merge in other valid `Partial<WidgetState>` objects, then the result will be a `WidgetState`.

Prior to this RFC (the current state), there is _no_ type checking that the arguments to `mergeProps` are valid `WidgetState` objects, and it just gets cast to `WidgetState` at the end, hoping for the best. The goal of this RFC is to improve type checking for that, and also to standardize how we define `WidgetState` types based on `WidgetProps`.

--

To your question about whether it's a public API: yes, `useWidget` is a public API that can be used by app developers to customize/extend the `Widget` component.

--

And shorthandProps is an array because that's what `makeMergeProps` and `resolveShorthandProps` expect(?). I'm trying to improve type safety without completely redesigning the APIs, and this pattern already existed (just with less type safety).